dcos_test_utils/gce.py

```diff
@@ -0,0 +1,254 @@
+import logging
+import typing
+from functools import wraps
+
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+from oauth2client.client import GoogleCredentials
+from retrying import retry
+
+from dcos_test_utils.helpers import Host
+
+log = logging.getLogger(__name__)
+
+# mapping used for the commonly used os name formats that differ from their respective formats in gce.
+# Update these mappings to expand OS support
+OS_IMAGE_FAMILIES = {
+    'cent-os-7': 'centos-7',
+    'ubuntu-16-04': 'ubuntu-1604-lts',
+    'coreos': 'coreos-stable',
+}
+
+# used in the gce sourceImage link (instance template field)
+IMAGE_PROJECTS = {
+    'centos': 'centos-cloud',
+    'rhel': 'rhel-cloud',
+    'ubuntu': 'ubuntu-os-cloud',
+    'coreos': 'coreos-cloud',
+    'debian': 'debian-cloud'
+}
+
+# template for an "instance template" resource to be used in a managed instance group
+INSTANCE_TEMPLATE = """
+- type: compute.v1.instanceTemplate
+  name: {name}
+  metadata:
+    dependsOn:
+    - {network}
+  properties:
+    project: {project}
+    properties:
+      machineType: {machineType}
+      disks:
+      - deviceName: boot
+        type: PERSISTENT
+        boot: true
+        autoDelete: true
+        initializeParams:
+          sourceImage: projects/{imageProject}/global/images/family/{sourceImage}
+      networkInterfaces:
+      - network: global/networks/{network}
+        # Access Config required to give the instance a public IP address
+        accessConfigs:
+        - name: External NAT
+          type: ONE_TO_ONE_NAT
+      metadata:
+        items:
+        - key: ssh-keys
+          value: {ssh_user}:{ssh_public_key}"""
+
+# template for a network resource in a gce deployment
+NETWORK_TEMPLATE = """
+- type: compute.v1.network
+  name: {name}
+  properties:
+    autoCreateSubnetworks: True"""
+
+# template for an instance group manager resource in a gce deployment
+MANAGED_INSTANCE_GROUP_TEMPLATE = """
+- type: compute.v1.instanceGroupManager
+  name: {name}
+  metadata:
+    dependsOn:
+    - {instance_template_name}
+  properties:
+    baseInstanceName: vm
+    instanceTemplate: global/instanceTemplates/{instance_template_name}
+    zone: {zone}
+    targetSize: {size}"""
+
+# template for a firewall in the network of a gce deployment
+FIREWALL_TEMPLATE = """
+- type: compute.v1.firewall
+  name: {name}
+  metadata:
+    dependsOn:
+    - {network}
+  properties:
+    description: allow all ports
+    network: global/networks/{network}
+    sourceRanges:
+    - 0.0.0.0/0
+    allowed:
+    - IPProtocol: tcp
+    - IPProtocol: udp
+    - IPProtocol: icmp
+    - IPProtocol: sctp"""
```

Comment:
This would break the deployment config since there can't be a new line in-between resources, but the superstrings will all be changed to dictionaries so this won't be a problem anymore